[SIP-5] Open a new /api/v1/query endpoint that takes query_obj

xtinec opened the pull request:
- Introduce a new handle_superset_exception decorator to avoid repeating the logic for catching SupersetExceptions
- Create a new /api/v1/query endpoint that takes a query_context which contains datasource info and the query_object. Note the query_context will be constructed in the client. The endpoint currently only initializes corresponding query_context and query_object based on data it receives. It doesn't have an implementation for `get_data`.
- Unit test to verify the new endpoint throws an exception.

@mistercrunch @kristw @williaster @john-bodley @conglei

Commit:
[SIP-5] Open a new /api/v1/query endpoint that takes query_obj

- Introduce a new handle_superset_exception decorator to avoid repeating the logic for catching SupersetExceptions
- Create a query_obj_backfill method that takes form_data and constructs a query_obj that will be constructed in the client in the future. Use the backfill in explore_json.
- Create a new /api/v1/query endpoint that takes query_obj only and returns the payload data. Note the query_obj is constructed in the client. The endpoint currently only handles query_obj for table view viz (we'll be adding support to new viz types as we go).
- Unit test to verify the new endpoint for table view

Commit:
fix tests and lint errors

kristw reviewed (commented):
LGTM. I'm not python expert tho.

mistercrunch commented on superset/models/core.py:
```diff
@@ -87,6 +87,37 @@ def copy_dashboard(mapper, connection, target):
+            'id': datasource_id,
+            'type': datasource_type,
+        }
+    return query_obj
```
Also it could be nice to have a class here for the query object, that way the have method like `get_cache_key`, `get_results_and_cache`, ...

xtinec replied:
The `q_obj` field of the query_obj is the real query obj that matches the interface for the connector. It also contains other fields that the viz classes need to initialize. 

I like the suggestion of building a class for these. We're definitely overloading the term query obj. Perhaps, we should call the class that contains both the real query obj and other things the viz views need query_payload or query_request or something similar?

williaster replied:
^like all of these ideas 👍

mistercrunch replied:
I don't feel strongly about names, but how about "QueryContext" ?

xtinec replied:
I like it! `QueryContext` it is. :)

mistercrunch commented on superset/utils/core.py:
```diff
@@ -213,6 +213,29 @@ def dttm_from_timtuple(d):
         d.tm_year, d.tm_mon, d.tm_mday, d.tm_hour, d.tm_min, d.tm_sec)
 
 
+def decode_iso_dttm(o):
```
Mmmh, can we just decode the ISOs values under the keys we know may be ISOs? If we want to cache to be determined from the query_obj, we'd need the relative time expressions here too, haven't given too much thought to this, but we should make a conscious decision as to whether query_obj would need to know about relative time exprs.

mistercrunch commented on superset/views/core.py:
```diff
@@ -1200,17 +1186,30 @@ def annotation_json(self, layer_id):
+    @api
+    @handle_superset_exception
+    @has_access_api
+    @expose('/api/v1/query/', methods=['POST'])
```
Would be nice to do this in a different file under `views/api.py` maybe

xtinec replied:
Yep, can do.

xtinec replied:
Yeah, I did a quick look around as well for how much needs to be moved to have a proper `api` folder under `app`. It is a non-trivial amount and probably not a priority to tackle right now. 

I'll stick with  `{ROOT}/views/api.py` for now.

mistercrunch reviewed (changes requested):
Closing a quick first round, happy to chat about all this.

xtinec commented on superset/models/core.py:
```diff
@@ -87,6 +87,37 @@ def copy_dashboard(mapper, connection, target):
+    # restructured as we build out the query object on the client side.
+    # The fields included are necessary to load a table viz without form_data
+    query_obj = {
+        'viz_type': form_data.get('viz_type', 'table'),
```
hmm, the downstream method, `get_viz` (which `get_payload` is called from) needs to know the `viz_type` in order to initialize, so we'll have to provide `viz_type` somehow. We could pass in `viz_type` as a parameter to `/api/v1/query`. Thoughts?

williaster replied:
I agree it should be `viz_type` because otherwise charts will be coupled with the backend, which is what we're trying to move away from for the `ChartPlugin`s.

conglei replied:
Eventually, in viz.py (if we still keep it), we will only have the BaseVis class. So here, actually we can first try the visualization type WorldCloud and BigNumber, since the only thing left for this two viz classes is query object logic.

xtinec commented on superset/models/core.py:
```diff
@@ -87,6 +87,37 @@ def copy_dashboard(mapper, connection, target):
+        'viz_type': form_data.get('viz_type', 'table'),
+        'time_range': form_data.get('time_range'),
+        'include_time': form_data.get('include_time'),
+        'token': form_data.get('token'),
```
This is a port of the token generation code from 
https://github.com/apache/incubator-superset/blob/master/superset/viz.py#L76-L77

It is being used in generating the `json_data` property for a viz
https://github.com/apache/incubator-superset/blob/master/superset/viz.py#L469-L490

The fields used to calculate cache keys looks to be something else entirely
https://github.com/apache/incubator-superset/blob/master/superset/viz.py#L338-L360

No idea what this token field is intended for. Any thoughts?

williaster replied:
@graceguo-supercat @michellethomas @john-bodley do any of you know?

mistercrunch replied:
I think a long time ago we used this to know whether the data came from cache or not and would define whether we needed to re-render the viz or not. I greped around and it looks unused at this point. Let's delete it.

xtinec replied:
👍

mistercrunch commented on superset/models/core.py:
```diff
@@ -85,6 +85,37 @@ def copy_dashboard(mapper, connection, target):
     session.commit()
 
 
+def query_obj_backfill(form_data, datasource_id=None, datasource_type=None):
```
There was a time where `form_data` did not include datasource_id & datasource_type, but now it always does in the shape of `fd.get('datasource') == '231__sqla'` or something like that.

xtinec replied:
Good to know! I'll remove them, which should clean up the signature of this method.

mistercrunch commented on superset/views/core.py:
```diff
@@ -1109,29 +1116,27 @@ def get_samples(self, viz_obj):
 
+    def check_datasource_permission(self, datasource, user=None):
+        if not security_manager.datasource_access(datasource, user):
+            return json_error_response(
```
Shouldn't this raise instead of returning?

xtinec replied:
I was returning an error json previously. I can change this to a `raise` for the new `handle_superset_exception` decorator to catch and handle.

mistercrunch commented on superset/views/core.py:
```diff
@@ -1109,29 +1116,27 @@ def get_samples(self, viz_obj):
+            form_data=form_data,
+            force=force,
+        )
+        self.check_datasource_permission(viz_obj.datasource, g.user)
```
check_datasource_permission seems like it returns a json error message, and it looks like here it assumes that it raises.

xtinec replied:
Good catch. Fixing.

mistercrunch commented on superset/views/core.py:
```diff
@@ -1200,17 +1186,30 @@ def annotation_json(self, layer_id):
+        query_obj = json.loads(request.form.get('query_obj'),
+                               object_hook=utils.decode_iso_dttm)
+        viz_obj = self.get_viz(query_obj=query_obj)
+        self.check_datasource_permission(viz_obj.datasource, g.user)
```
Same as mentioned above, this looks like it assumes that `check_datasource_permission` raises

xtinec replied:
ditto

mistercrunch commented on superset/views/core.py:
```diff
@@ -1410,10 +1404,7 @@ def filter(self, datasource_type, datasource_id, column):
-            return json_error_response(
-                security_manager.get_datasource_access_error_msg(datasource))
-
+        self.check_datasource_permission(datasource)
```
Same as mentioned above, this looks like it assumes that check_datasource_permission raises

xtinec replied:
ack

mistercrunch commented on superset/views/core.py:
```diff
@@ -2643,10 +2634,7 @@ def fetch_datasource_metadata(self):
-            return json_error_response(
-                security_manager.get_datasource_access_error_msg(datasource),
-                link=security_manager.get_datasource_access_link(datasource))
+        self.check_datasource_permission(datasource)
```
Same as mentioned above, this looks like it assumes that check_datasource_permission raises

xtinec replied:
ditto

mistercrunch commented on superset/views/core.py:
```diff
@@ -2831,11 +2819,7 @@ def slice_query(self, slice_id):
-                security_manager.get_datasource_access_error_msg(viz_obj.datasource),
-                status=401,
-                link=security_manager.get_datasource_access_link(viz_obj.datasource))
+        self.check_datasource_permission(viz_obj.datasource)
```
Same as mentioned above, this looks like it assumes that check_datasource_permission raises

xtinec replied:
ditto

mistercrunch commented on superset/viz.py:
```diff
@@ -63,22 +65,42 @@ class BaseViz(object):
+                datasource_dict.get('type'), datasource_dict.get('id'), db.session)
+            if not ds:
+                raise Exception(_('Viz is missing a datasource'))
+            else:
```
NIT: no need for an else when the if clause exits

xtinec replied:
👍

mistercrunch commented on superset/viz.py:
```diff
@@ -63,22 +65,42 @@ class BaseViz(object):
-        self.query = ''
-        self.token = self.form_data.get(
-            'token', 'token_' + uuid.uuid4().hex[:8])
+        self.viz_type = query_obj.get('viz_type')
```
Do we need to unpack all of this? Could we just `self.query_obj = query_obj` ?

xtinec replied:
Not any more once we have `QueryContext`. :)

Commit:
- Move the new query endpoint into its own api.py view.
- Create QueryObject and QueryContext class to encapsulate query_object to be built from the client and additional info (e.g. datasource) needed to get the data payload for a given query
- Remove the query_obj_backfill as we'll start building the first query_object on the client so it no longer makes sense to have a short-lived backfill for the matter of days.

Commit:
Fixing lint and test errors

Commit:
Fixing additional lint error from the previous rebase.

xtinec committed:
Merge branch 'master' into query-obj-end-point

Commit:
fixing additional lint error

mistercrunch commented on superset/security.py:
```diff
@@ -428,3 +429,10 @@ def set_perm(self, mapper, connection, target):  # noqa
                 ),
             )
+
+    def check_datasource_permission(self, datasource, user=None):
```
👍 nice refactor, tiny NIT where the name `assert_datasource_permission ` could be a bit more clear about the fact that it's a "raise if that's not the case". `check_%` could sound more like it's returning a boolean/

xtinec replied:
👍

mistercrunch commented on superset/views/base.py:
```diff
@@ -75,6 +84,26 @@ def wraps(self, *args, **kwargs):
     return functools.update_wrapper(wraps, f)
 
 
+def handle_superset_exception(f):
```
👍

williaster commented on superset/common/query_object.py:
```diff
@@ -0,0 +1,59 @@
+            since: Optional[str] = None,
+            until: Optional[str] = None,
+            time_shift: Optional[str] = None,
+            is_timeseries: bool = False,
```
do we need the `is_timeseries` flag?

xtinec replied:
Yeah, I think so. We're currently setting it to different values depending on which viz it is on the server side. My thought is we'll compute this on the client and pass it in as part of the `query_object`. Thoughts?

conglei replied:
yes, we do need it, and it is one field of the query_object.

williaster replied:
@conglei long-term is the plan to remove it?

williaster commented on superset/common/query_object.py:
```diff
@@ -0,0 +1,59 @@
+            groupby: List[str],
+            metrics: List[Metric],
+            filters: List[str],
+            time_range: Optional[str] = None,
```
do we need to support `time_range` _and_ `since` + `until`? in an ideal world it'd be nice to not have time ranges coming from two different places.

xtinec replied:
Good question! 

I agree we should only have one. Looking at the code, seems like the calculation based on `time_range` encompass the one based on `since` and `until`. I'll have the new `query_object` only take `time_range`. 😄

williaster replied:
sweet!

xtinec commented on superset/common/query_context.py:
```diff
@@ -0,0 +1,31 @@
+        self._query_object = QueryObject(**query_object)
+
+    @property
+    def datasource(self):
```
👍 Updating.

xtinec commented on superset/common/query_object.py:
```diff
@@ -0,0 +1,59 @@
+        pass
+
+    def get_data(self):
+        # TODO: implement
```
I could change this to `raise NotImplementedError()` here and verify in the new unit test the exception is raised.

xtinec commented on superset/connectors/connector_registry.py:
```diff
@@ -18,6 +18,8 @@ def register_sources(cls, datasource_config):
     @classmethod
     def get_datasource(cls, datasource_type, datasource_id, session):
+        print('session')
+        print(session)
```
😅

xtinec commented on superset/utils/core.py:
```diff
@@ -271,7 +272,7 @@ def parse_human_timedelta(s):
     cal = parsedatetime.Calendar()
     dttm = dttm_from_timtuple(datetime.now().timetuple())
-    d = cal.parse(s, dttm)[0]
+    d = cal.parse(s or '', dttm)[0]
```
`cal.parse(None)` throws a `TypeError`. To stay true to the existing implementation (return a delta when `None` is passed in), we have to do `s or ''`, I think.

xtinec commented on superset/utils/core.py:
```diff
@@ -940,15 +940,20 @@ def get_since_until(form_data):
 
+def check_from_to_dttm(from_dttm, to_dttm):
+    if from_dttm and to_dttm and from_dttm > to_dttm:
+        raise Exception(_('From date cannot be larger than to date'))
```
👍

Commit:
addressing additional pr comments

conglei commented on superset/common/query_object.py:
```diff
@@ -0,0 +1,56 @@
+            metrics: List[Metric],
+            filters: List[str],
+            time_range: Optional[str] = None,
+            time_shift: Optional[str] = None,
```
I don't think we need this though, since we will handle the time_shift in the client.

xtinec replied:
hmm, this should probably be handled on the server like `time_range`. The currently dttm parsing logic is quite complex (see below) and we run the risk of having time-zone related discrepancies if we parse timestamps on the client. 

```
since_dttm, until_dttm = utils.get_since_until(time_range)
time_shift_dttm = utils.parse_human_timedelta(time_shift)
self.from_dttm = None if since_dttm is None else (since_dttm - time_shift_dttm)
self.to_dttm = None if until_dttm is None else (until_dttm - time_shift_dttm)
```
Thoughts?

conglei replied:
You are right! this is the correct way to do that. Didn't think about the calculation part.

conglei commented on superset/common/query_context.py:
```diff
@@ -0,0 +1,24 @@
+        self.datasource = ConnectorRegistry.get_datasource(datasource.get('type'),
+                                                           datasource.get('id'),
+                                                           db.session)
+        self.query_object = QueryObject(**query_object)
```
As we discussed, the query_object should be an array, or should we make it as an array of `QueryObject`?

xtinec replied:
Yep, I can change this to an array of `QueryObject`s

conglei commented on superset/views/api.py:
```diff
@@ -0,0 +1,32 @@
+        query_context = QueryContext(**json.loads(request.form.get('query_context')))
+        print(query_context)
+        security_manager.assert_datasource_permission(query_context.datasource, g.user)
+        payload_json = query_context.query_object.get_data()
```
Since now `query_object` won't have the datasource info, we may need to move `get_data` as a function in `query_context`

conglei replied:
but I think I can revise it later in my PR.

xtinec replied:
👍

xtinec replied:
Actually, it is trivial to move the empty implementation, so I moved the method into `QueryContext`. 😄

Commit:
Make /query accept a list of queries in the query_context object.

Commit:
fixing a lint error

mistercrunch commented on superset/utils/core.py:
```diff
@@ -940,15 +940,20 @@ def get_since_until(form_data):
     return since, until
 
 
+def assert_from_to_dttm(from_dttm, to_dttm):
```
NIT (optional): to me there's no need to refactor out a 2 line methods into the `utils/` package if it's used in a single place in the code base. At that point it's just unnecessary indirection.

xtinec replied:
What do you think of computing `from_dttm` and `to_dttm` and asserting their validity as part of `get_since_util` (i.e. moving the logic in the blob below into `get_since_util`). We'll provide time_shift as an additional argument to get_since_until. 
```
        self.from_dttm = None if since_dttm is None else (since_dttm - time_shift_dttm)
        self.to_dttm = None if until_dttm is None else (until_dttm - time_shift_dttm)
        utils.assert_from_to_dttm(self.from_dttm, self.to_dttm)
```

My thought is all these time computation and validation should be 1) reused (it is not currently doesn't mean it can't be) and 2) kept out of the main constructor of `QueryObject` to make the constructor code easy to read.

Otherwise, I'm happy to put the two lines back.

mistercrunch replied:
I don't feel strongly about indirections but raising on the `get_since_util` SGTM

xtinec replied:
Moving them to all to utils and adding types. I quite like it now. 😄 

```
if time_shift:
        time_shift = parse_human_timedelta(time_shift)
        since = since if since is None else (since - time_shift)
        until = until if until is None else (until - time_shift)

    if since and until and since > until:
        raise ValueError(_('From date cannot be larger than to date'))

    return since, until
```
Then, in the constructor, we simply have
``` 
self.from_dttm, self.to_dttm = utils.get_since_until(time_range, time_shift)
```
Updates coming shortly.

Commit:
- Move time_shift based calculation and since, until check into util
- Add typing info for get_since_until
- Add new unit tests to verify time_shift calculation and the since until check

mistercrunch commented on superset/views/core.py:
```diff
@@ -1118,29 +1114,19 @@ def get_samples(self, viz_obj):
             'data': viz_obj.get_samples(),
         })
 
+    @handle_superset_exception
```
Just rebased and trying to fix stacktraces, and I think applying the decorator here is wrong since this method is not a view
